File: MilestoneThree/CS499_ArtifactTwo_Original/ItemTracker.cpp
#include "ItemTracker.h"
#include <iostream>
#include <fstream>

using namespace std;

ItemTracker::ItemTracker() { // Constructor
    itemFrequency = map<string, int>(); // Initialize map
}

void ItemTracker::ReadFile(string filename) { // Read .txt file
   ifstream inputFile(filename); // Open .txt file
    if (!inputFile) { // Check if file sucessfully opened
        cout << "Could not open file: " << filename << endl;
        return;
    }
   
    string item; 

    while (inputFile >> item) { // Loop through each item in the file
        itemFrequency[item]++;// Increment item frequency
    }
    inputFile.close();// Close the file
}

int ItemTracker::GetItemFrequency(const string& item) { // Get item frequency
    if (itemFrequency.count(item) > 0) { // Check for item in map
        return itemFrequency.at(item); // Return frequency
    }
    return 0;
}

void ItemTracker::PrintItemFrequency() { // Print item frequency
    if (itemFrequency.empty()) { // Check if empty
        cout << "No items available." << endl;
        return;
    }
    for (pair<string, int> item : itemFrequency) {// Loop through map and print item with frequency
        cout << item.first << " " << item.second << endl;
    }
}

void ItemTracker::PrintHistogram() { // Print histogram
    if (itemFrequency.empty()) { // Check if empty 
        cout << "Item not available." << endl; 
        return;
    }
    for (pair<string, int> item: itemFrequency) { // Loop through map 
        cout << item.first << " ";
        for (int i = 0; i < item.second; i++) {// Print "*" matching the item frequency
            cout << "*";
        }
        cout << endl;
    }
}

void ItemTracker::SaveFile(const string& filename) { // Save file to frequency.dat
    ofstream outputFile(filename);
    if (!outputFile) { // Check if file successfully opened
        cout << "Could not open file: " << filename << endl;
        return;
    }
    for (pair<string, int> item : itemFrequency) {// Loop through map and write to file
        outputFile << item.first << " " << item.second << endl;
    }
    outputFile.close(); // Close the file
}

File: MilestoneTwo/CS499_ArtifactOne_Enhancement/Item_Tracker.py
"""
CS499 Capstone Artifact One

Original Project: 
C++ item tracker

Enhancements:
Convert from c++ to Python while maintaining OOP and improving code.
"""

class ItemTracker:
    # Initialize dictionary for key value storage for the item frequency
    def __init__(self):
        self.item_frequency = {}

    # Read items and stores frequency from the file 
    def read_file(self, file_name):
        try:
             with open(file_name, 'r') as input_file:
                  for item in input_file:
                       # Normalized input 
                       item = item.strip().lower()

                       self.item_frequency[item] = self.item_frequency.get(item, 0) + 1
        except FileNotFoundError:
            print(f"Could not open file: {file_name}")

    # Get the item frequency of the searched item
    def get_item_frequency(self, item):
        item = item.strip().lower()
        return self.item_frequency.get(item, 0)
    
    # Print all items with their frequencies
    def print_item_frequency(self):
        if not self.item_frequency:
            print("No items available.")
            return
        for item, frequency in self.item_frequency.items():
            print(f"{item.title()} {frequency}")

    # Print each item with a histogram with asterisks
    def print_histogram(self):
        if not self.item_frequency:
            print("Item not available.")
            return
        for item, frequency in self.item_frequency.items():
            print(f"{item.title()} {'*' * frequency}")

    # Save item frequency data to output file (frequency.dat)
    def save_file(self, file_name):
        try:
            with open(file_name, 'w') as output_file:
                for item, frequency in self.item_frequency.items():
                    output_file.write(f"{item.title()} {frequency}\n")
        except Exception as exception:
            print(f"Could not open file: {file_name}")
            return
        
def display_menu():
    # Display a menu for the user
    print("|=======================================|")
    print("Please choose an option from the menu:")
    print("1. Get item frequency.")
    print("2. Print list of items and the frequency.")
    print("3. Print a histogram of items.")
    print("4. Exit The Program!")
    print("|=======================================|")
    
# Main
def main():
        # Create object, input file, and read and save frequency data
        item_tracker = ItemTracker()
        file_name = "CS210_Project_Three_Input_File.txt"
        item_tracker.read_file(file_name)
        item_tracker.save_file("frequency.dat")
        choice = 0

        while True:
            display_menu()

            # Validate the user input
            try: 
                choice = int(input("Enter a choice from the menu: "))
            except ValueError:
                print("Invalid input. Please choose 1 - 4 from the menu options")
                continue

            # Menu selection
            if choice == 1:
                item = input("Enter the item name: ")
                frequency = item_tracker.get_item_frequency(item)
                print(f"The frequency of {item.title()} are {frequency}")
            elif choice == 2:
                item_tracker.print_item_frequency()
            elif choice == 3:
                 item_tracker.print_histogram()
            elif choice == 4:
                print("Exit The Program")
                break
            else: 
                print("Invalid choice.")

# Start program
if __name__ == "__main__":
    main()  